sycl/source/ld-version-script.txt

```diff
@@ -0,0 +1,25 @@
+{
+  global:
+    /* Export everything from cl::sycl namespace */
+    _ZNK2cl4sycl*;  /* function */
```

Comment:
I think it's better to use C++ language instead of C. Is there any problems with that?

Reply:
Described in the commit message:
> This patch does not use extern "C++" matcher for C++ functions,
> because vtable and typeinfo symbols make extern "C++" patterns more
> complicated than patterns against mangled names.
>
> With extern "C++" we have to match for "vtable for cl::sycl::foo", but
> not match for "vtable for std::__internalcl::sycl::foo".

Basically, this way it is just easier to read and understand.

Reply:
> > This patch does not use extern "C++" matcher for C++ functions,
> > because vtable and typeinfo symbols make extern "C++" patterns more
> > complicated than patterns against mangled names.
> > With extern "C++" we have to match for "vtable for cl::sycl::foo", but
> > not match for "vtable for std::__internalcl::sycl::foo".
> 
> Basically, this way it is just easier to read and understand.

Will this work?
```
vtable for cl::sycl::*
typeinfo for cl::sycl::*
```
This seems looks quite readable.

Reply:
> > > This patch does not use extern "C++" matcher for C++ functions,
> > > because vtable and typeinfo symbols make extern "C++" patterns more
> > > complicated than patterns against mangled names.
> > > With extern "C++" we have to match for "vtable for cl::sycl::foo", but
> > > not match for "vtable for std::__internalcl::sycl::foo".
> > 
> > 
> > Basically, this way it is just easier to read and understand.
> 
> Will this work?
> 
> ```
> vtable for cl::sycl::*
> typeinfo for cl::sycl::*
> ```
> 
> This seems looks quite readable.

IIRC, spaces are not allowed without quotes, and a quoted string cannot have a glob.

Reply:
If I understand correctly, spaces can be replaced with `?`. I.e.
```
vtable?for?cl::sycl::*;
typeinfo?for?cl::sycl::*;
```

Reply:
Probably.
If you believe that this is more readable (looks the same for me) and it is important, then I can use this pattern for C++ matching.

Reply:
I think using C++ language is intended for C++ libraries, which is our case.
https://sourceware.org/binutils/docs/ld/VERSION.html

Reply:
It is just a helper that demangle a symbol name. It is not required.

> You **can** also specify the language in the version script:
> The linker will iterate over the list of symbols at the link time and demangle them according to ‘lang’ before matching them to the patterns

Reply:
While I find the C++ clearer, it seems more resilient to some future change in the mangling.

Also please distribute the useful description of the PR as comments in the PR itself, otherwise this design information will be lost forever... :-(

Reply:
Added a note about extern "C++".
The overall problem should probably be described in the ABI guidelines document (https://github.com/intel/llvm/pull/907#issuecomment-562229842). 

@alexbatashev, FYI.

Reply:
@keryell, I think @asavonic is confused by your comment.
According to my interpretation, you are in favor of using extern "C++". Please, confirm.

@asavonic, I still think that using extern "C++" is better option here.

Reply:
extern "C++" is also inconvenient when demangler puts a return type before the function name:
```
_ZNK2cl4sycl6device18create_sub_devicesILNS0_4info18partition_propertyE4231EEESt6vectorIS1_SaIS1_EERKS5_ImSaImEE
```
gets demangled into
```
std::vector<cl::sycl::device, std::allocator<cl::sycl::device> > cl::sycl::device::create_sub_devices<(cl::sycl::info::partition_property)4231>(std::vector<unsigned long, std::allocator<unsigned long> > const&) const
```

I do not want to mess with demangled names. As I already stated, it complicates things for no clear benefit.

Reply:
Yes my sentence was unclear. Sorry. I was in favor of C++ for 2 reasons but the current solution is also fine if it works.